### PR TITLE
telemetry: Add `language_name` and `model_provider`

### PR DESCRIPTION
**Pull request by bcherny**
This PR adds a bit more metadata for assistant logging.

Release Notes:

- Assistant: Added `language_name` and `model_provider` fields to telemetry events.

### CODE REVIEW

**Comment by bcherny** [comment]
@cla-bot check

**Review by rgbkrk — APPROVED** [review]

File: crates/assistant/src/context.rs [hunk 1]
```diff
@@ -2133,14 +2133,17 @@ impl Context {
                     });
 
                     if let Some(telemetry) = this.telemetry.as_ref() {
-                        telemetry.report_assistant_event(
```
Comment (bcherny) [inline crates/assistant/src/context.rs:2136]: Ran into the Clippy 7-arg limit when I just added to this

File: crates/client/src/telemetry.rs [hunk 2]
```diff
@@ -536,7 +541,7 @@ impl Telemetry {
         self.report_event(event)
     }
 
-    pub fn report_event(self: &Arc<Self>, event: Event) {
+    fn report_event(self: &Arc<Self>, event: Event) {
```
Comment (rgbkrk) [inline crates/client/src/telemetry.rs:544]: 👍

File: crates/multi_buffer/src/multi_buffer.rs [hunk 3]
```diff
@@ -382,6 +382,15 @@ impl MultiBuffer {
         }
     }
 
+    pub fn language<'a>(&self, cx: &'a AppContext) -> Option<&'a Arc<Language>> {
+        // Take the last language, since all buffer are guaranteed to be
```
Comment (maxdeviant) [inline crates/multi_buffer/src/multi_buffer.rs:386]: I don't think this is true?

What is guaranteeing that each buffer in the multibuffer are in the same file?

**Review by maxdeviant — CHANGES_REQUESTED** [review]
How we're determining the language for a multibuffer doesn't seem correct.

File: crates/multi_buffer/src/multi_buffer.rs (hunk 3 above)
Comment (bcherny) [reply crates/multi_buffer/src/multi_buffer.rs:386]: I don't believe there is a way to open an inline assistant across files, right? All the buffers must be in the same file from what I could tell.

Comment (maxdeviant) [reply crates/multi_buffer/src/multi_buffer.rs:386]: Even if it were, that property is not true for the multibuffer as a whole, so having this `language` getter available for every multibuffer is not correct.

We'd need to do something that is specific to the inline assist case.

> I don't believe there is a way to open an inline assistant across files, right?

It is possible to have multiple inline assists open in a multibuffer that spans multiple files. Here's an example where I initiated an inline assist from within a project search:

Screenshot 2024-10-04 at 1 38 55 PM

Comment (bcherny) [reply crates/multi_buffer/src/multi_buffer.rs:386]: Gotcha. Any suggestions for how to grab the correct buffer for a given inline assistant? If that's hard to do, we can also send up a Vec of languagenames.

Comment (maxdeviant) [reply crates/multi_buffer/src/multi_buffer.rs]: Max and I are taking a look at it now.

Comment (maxdeviant) [reply crates/multi_buffer/src/multi_buffer.rs]: We removed the `language` method in the multibuffer and addressed inferring the language for the assist on a case-by-case basis.

We're able to use the range of the inline assistant to work out which buffer we're in and get that language's multibuffer:

https://github.com/zed-industries/zed/pull/18640/commits/12b95cedcda5be542e547c6b3079a863f505241f

Comment (bcherny) [reply crates/multi_buffer/src/multi_buffer.rs]: Oh that looks solid! Thank you for jumping in.

**Review by maxdeviant — APPROVED** [review]
Thanks!

**Pull request merged**